momaland/envs/gem_mining/gem_mining.py

```diff
@@ -0,0 +1,274 @@
+        self.agents = self.possible_agents[:]
+        self.num_mines = num_agents + max_connectivity - 1
+
+        self.num_objectives = num_objectives
+        self.worker_bonus = w_bonus
+        self.truncation_probability = trunc_probability
+
+        self.random = random.Random(seed)
+        self.np_random = np.random.RandomState(seed)
+
+        # determine the number of workers per village (agent):
+        lst = list(range(min_workers, max_workers + 1))
+        print(lst)
+        self.workers = self.random.choices(lst, k=self.num_agents)
+
+        # determine the base probabilities of finding a gem per type per mine
+        self.base_probabilities = dict()
+        for i in range(self.num_mines):
+            self.base_probabilities[i] = [0] * self.num_objectives
+            left = max_prob - self.num_objectives * min_prob
+            for j in range(self.num_objectives):
+                if correlated_objectives:
+                    pj = self.random.uniform(min_prob, left)
+                    left = left - pj
+                else:
+                    pj = self.random.uniform(min_prob, max_prob)
+                self.base_probabilities[i][j] = pj
+            self.random.shuffle(self.base_probabilities[i])
+
+        # action spaces are numbers (IDs) of the mines where each agent can go
+        self.action_spaces = dict()
+        for i in range(num_agents):
+            connect = self.random.randint(min_connectivity, max_connectivity)
+            self.action_spaces[i] = Discrete(connect, start=i)
+        # stateless setting, agents receive a constant '0' as an observation in each timestep
+        self.observation_spaces = dict(
+            zip(
+                self.agents,
+                [
+                    Box(
+                        low=0,
+                        high=self.num_agents,
+                        shape=(1,),
+                        dtype=np.float32,
+                    )
+                ]
+                * num_agents,
+            )
+        )
+        self.reward_spaces = dict(zip(self.agents, [Box(low=0, high=self.num_mines, shape=(num_objectives,))] * num_agents))
+
+        # set truncations to false (no agents are ever lost)
+        self.truncations = {agent: False for agent in self.agents}
+
+    # this cache ensures that same space object is returned for the same agent
+    # allows action space seeding to work as expected
+    @functools.lru_cache(maxsize=None)
+    @override
+    def observation_space(self, agent):
+        return self.observation_spaces[agent]
+
+    @functools.lru_cache(maxsize=None)
+    @override
+    def action_space(self, agent):
+        return self.action_spaces[agent]
+
+    @override
+    def reward_space(self, agent):
+        return self.reward_spaces[agent]
+
+    @override
+    def render(self):
+        if self.render_mode is None:
+            warn("You are calling render method without specifying any render mode.")
+            return
```

Comment:
Am I correct in understanding that render either gives a warning or does nothing? The reason I'm asking is because render modes "human" was said to be supported but it appears not to be the case.